File: landrop-main/receiver/android/app/src/main/java/com/landrop/receiver/protocol/Validation.kt
package com.landrop.receiver.protocol

class ProtocolProblem(val code: ErrorCode, val field: String) :
    IllegalArgumentException("${code.name}: $field")

// Configurable implementation defaults, not normative specification limits.
// All text lengths are UTF-8 bytes, matching Rust rather than UTF-16 units.
data class Limits(
    val maxJsonBytes: Int = 65536,
    val maxTextBytes: Int = 256,
    val maxDescriptionBytes: Int = 4096,
    val maxApkBytes: Long = 4_294_967_296,
)

internal fun invalid(field: String): Nothing = throw ProtocolProblem(ErrorCode.INVALID_REQUEST, field)
private fun metadata(field: String): Nothing = throw ProtocolProblem(ErrorCode.INVALID_METADATA, field)
internal fun wellFormedUnicode(value: String): Boolean {
    var i = 0
    while (i < value.length) {
        val c = value[i++]
        if (c.isHighSurrogate()) {
            if (i == value.length || !value[i++].isLowSurrogate()) return false
        } else if (c.isLowSurrogate()) return false
    }
    return true
}
private fun text(value: String, max: Int, empty: Boolean, field: String) {
    val blank = value.trim(' ', '\t', '\r', '\n').isEmpty()
    if (!wellFormedUnicode(value) || value.toByteArray(Charsets.UTF_8).size > max ||
        '\u0000' in value || (!empty && blank)) metadata(field)
}
fun ProtocolVersion.validate(limits: Limits = Limits()) {
    if (value <= 0) invalid("protocol_version")
    if (value != CURRENT_PROTOCOL_VERSION) throw ProtocolProblem(ErrorCode.UNSUPPORTED_PROTOCOL, "protocol_version")
}
fun SenderProfile.validate(limits: Limits = Limits()) {
    text(displayName, limits.maxTextBytes, false, "sender.display_name")
}
fun ApplicationMetadata.validate(limits: Limits = Limits()) {
    val packagePattern = Regex("[A-Za-z][A-Za-z0-9_]*(\\.[A-Za-z][A-Za-z0-9_]*)+")
    if (identity.packageName.length > 255 || !packagePattern.matches(identity.packageName)) metadata("application.package_name")
    if (identity.sizeBytes <= 0) metadata("application.size_bytes")
    if (identity.sizeBytes > limits.maxApkBytes) throw ProtocolProblem(ErrorCode.FILE_TOO_LARGE, "application.size_bytes")
    text(presentation.name, limits.maxTextBytes, false, "application.name")
    text(presentation.version, limits.maxTextBytes, false, "application.version")
    text(presentation.description, limits.maxDescriptionBytes, true, "application.description")
}
fun ControlMessage.validate(limits: Limits = Limits()) {
    when (this) {
        is DeliveryRequest -> {
            protocolVersion.validate(limits)
            sender.validate(limits)
            application.validate(limits)
        }
        is DeliveryResponse -> Unit
        is ErrorMessage -> text(message, limits.maxDescriptionBytes, false, "message")
        is TransferProgress -> {
            // transfer_id validity is enforced by the TransferId value class at
            // construction. Byte counters must be non-negative and consistent
            // (bytesTransferred <= totalBytes), matching the TypeScript codec.
            if (bytesTransferred < 0 || totalBytes < 0 || bytesTransferred > totalBytes) {
                invalid("bytes_transferred")
            }
        }
        is TransferCancel -> Unit
    }
}
fun Lifetime.validate(limits: Limits = Limits()) {
    if (createdAtMs < 0 || expiresAtMs <= createdAtMs) invalid("lifetime")
}
fun ApkVersion.validate(limits: Limits = Limits()) {
    if (versionCode != null && versionCode < 0) metadata("version_code")
    versionName?.let { text(it, limits.maxTextBytes, true, "version_name") }
}
fun SessionData.validate(limits: Limits = Limits()) {
    protocolVersion.validate(limits)
    sender.validate(limits)
    text(receiver.deviceName, limits.maxTextBytes, false, "receiver.device_name")
    lifetime.validate(limits)
}

File: landrop-main/sender/desktop/src-tauri/src/protocol/validation.rs
use super::{ApkVersion, ApplicationMetadata, ControlMessage, DeliveryRequest, ErrorCode, Lifetime, ProtocolVersion, SenderProfile, SessionData, TransferProgress, CURRENT_PROTOCOL_VERSION};
use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProtocolProblem { pub code: ErrorCode, pub field: &'static str }
impl ProtocolProblem {
    pub(crate) fn invalid(field: &'static str) -> Self { Self { code: ErrorCode::InvalidRequest, field } }
}
impl fmt::Display for ProtocolProblem {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { write!(f, "{:?}: {}", self.code, self.field) }
}
impl std::error::Error for ProtocolProblem {}

// Configurable implementation defaults, not normative specification limits.
// Lengths are UTF-8 bytes, identically measured on Android.
#[derive(Debug, Clone)]
pub struct Limits {
    pub max_json_bytes: usize,
    pub max_text_bytes: usize,
    pub max_description_bytes: usize,
    pub max_apk_bytes: i64,
}
impl Default for Limits {
    fn default() -> Self {
        Self { max_json_bytes: 65536, max_text_bytes: 256, max_description_bytes: 4096, max_apk_bytes: 4_294_967_296 }
    }
}
pub trait Validate { fn validate(&self, limits: &Limits) -> Result<(), ProtocolProblem>; }
fn metadata(field: &'static str) -> ProtocolProblem { ProtocolProblem { code: ErrorCode::InvalidMetadata, field } }
fn text(value: &str, max: usize, empty: bool, field: &'static str) -> Result<(), ProtocolProblem> {
    let blank = value.trim_matches([' ', '\t', '\r', '\n']).is_empty();
    if value.len() > max || value.contains('\0') || (!empty && blank) { Err(metadata(field)) } else { Ok(()) }
}
impl Validate for ProtocolVersion {
    fn validate(&self, _: &Limits) -> Result<(), ProtocolProblem> {
        if self.0 <= 0 { Err(ProtocolProblem::invalid("protocol_version")) }
        else if self.0 != CURRENT_PROTOCOL_VERSION { Err(ProtocolProblem { code: ErrorCode::UnsupportedProtocol, field: "protocol_version" }) }
        else { Ok(()) }
    }
}
impl Validate for SenderProfile {
    fn validate(&self, l: &Limits) -> Result<(), ProtocolProblem> { text(&self.display_name, l.max_text_bytes, false, "sender.display_name") }
}
impl Validate for ApplicationMetadata {
    fn validate(&self, l: &Limits) -> Result<(), ProtocolProblem> {
        let package = &self.identity.package_name;
        let segments: Vec<_> = package.split('.').collect();
        let valid = package.len() <= 255 && segments.len() >= 2 && segments.iter().all(|s| {
            !s.is_empty() && s.as_bytes()[0].is_ascii_alphabetic()
                && s.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'_')
        });
        if !valid { return Err(metadata("application.package_name")); }
        if self.identity.size_bytes <= 0 { return Err(metadata("application.size_bytes")); }
        if self.identity.size_bytes > l.max_apk_bytes {
            return Err(ProtocolProblem { code: ErrorCode::FileTooLarge, field: "application.size_bytes" });
        }
        text(&self.presentation.name, l.max_text_bytes, false, "application.name")?;
        text(&self.presentation.version, l.max_text_bytes, false, "application.version")?;
        text(&self.presentation.description, l.max_description_bytes, true, "application.description")
    }
}
impl Validate for DeliveryRequest {
    fn validate(&self, l: &Limits) -> Result<(), ProtocolProblem> {
        self.protocol_version.validate(l)?;
        self.sender.validate(l)?;
        self.application.validate(l)
    }
}
impl Validate for ControlMessage {
    fn validate(&self, l: &Limits) -> Result<(), ProtocolProblem> {
        match self {
            Self::DeliveryRequest(v) => v.validate(l),
            Self::DeliveryResponse(_) => Ok(()),
            Self::Error(v) => text(&v.message, l.max_description_bytes, false, "message"),
            Self::TransferProgress(v) => v.validate(l),
            Self::TransferCancel(_) => Ok(()),
        }
    }
}
impl Validate for TransferProgress {
    fn validate(&self, _: &Limits) -> Result<(), ProtocolProblem> {
        // transfer_id validity is enforced by the TransferId newtype at
        // construction. Byte counters must be non-negative and consistent
        // (bytes_transferred <= total_bytes), matching the TypeScript codec.
        if self.bytes_transferred < 0 || self.total_bytes < 0 {
            return Err(ProtocolProblem::invalid("bytes_transferred"));
        }
        if self.bytes_transferred > self.total_bytes {
            return Err(ProtocolProblem::invalid("bytes_transferred"));
        }
        Ok(())
    }
}
impl Validate for Lifetime {
    fn validate(&self, _: &Limits) -> Result<(), ProtocolProblem> {
        if self.created_at_ms < 0 || self.expires_at_ms <= self.created_at_ms { Err(ProtocolProblem::invalid("lifetime")) } else { Ok(()) }
    }
}
impl Validate for ApkVersion {
    fn validate(&self, l: &Limits) -> Result<(), ProtocolProblem> {
        if self.version_code.is_some_and(|v| v < 0) { return Err(metadata("version_code")); }
        if let Some(v) = &self.version_name { text(v, l.max_text_bytes, true, "version_name")?; }
        Ok(())
    }
}
impl Validate for SessionData {
    fn validate(&self, l: &Limits) -> Result<(), ProtocolProblem> {
        self.protocol_version.validate(l)?;
        self.sender.validate(l)?;
        text(&self.receiver.device_name, l.max_text_bytes, false, "receiver.device_name")?;
        self.lifetime.validate(l)
    }
}
